### DISCUSSION TITLE
Limiting the number of attached models

### DISCUSSION THREAD

**Discussion by wimski** [laravel/nova-issues#5145] [Q&A]
I want to limit the number of models that can be attached to another model in Nova.

Let's say I have the following setup with `Foo` and `Bar`.

#### Models
```php
class Foo extends Model
{
    protected const MAXIMUM_BARS = 5;

    public function bars(): BelongsToMany
    {
        return $this->BelongsToMany(Bar::class);
    }

    public function hasMaximumBars(): bool
    {
        return $this->bars()->count() >= self::MAXIMUM_BARS;
    }
}

class Bar extends Model {}
```

#### Nova Resources
```php
class FooResource extends Resource
{
    public static string $model = Foo::class;

    public function fields(NovaRequest $request): array
    {
        return [
            BelongsToMany::make('Bars Label', 'bars', BarResource::class),
        ];
    }
}

class BarResource extends Resource
{
    public static string $model = Bar::class;
}
```

#### Policies
```php
class FooPolicy
{
    public function attachBar(User $user, Foo $foo, Bar $bar): bool
    {
        return ! $foo->hasMaximumBars();
    }

    public function attachAnyBar(User $user, Foo $foo): bool
    {
        return ! $foo->hasMaximumBars();
    }
}
```

* The `attachAnyBar` method hides the 'Attach Bar' button in the Foo detail view whenever the maximum is reached, although the `/attach/bar-resources` route is still accessible manually when I'd expect a 403.
* Once in the attach view, the button 'Attach & attach another' is always visible, no matter the policy. Only the found resources eligible for attachment seem to get limited to zero when the maximum is reached.

So it technically works in that a user is never able to attach more than the maximum, but this all feels like bad UX to me.

* Is there a special policy method specifically for this kind of situation that I'm not aware of?
* Is there a better way to limit the number of attachments?

**Comment by RealityGamin** [comment]
If you want a hack around it (which I don't necessarily like myself).. you can throw an abort error using an observer on barClass https://laravel.com/docs/9.x/eloquent#events

```
protected static function booted()
    {
        static::creating(function ($model) {
            // check if there is already 5 existing already, and throw abort error
        });
    }
```

Have you tried using the general attachAny rather than attachAnyBar to see if you will get the same result? you could also just directly do it from the resource to test it out

```
//On the Resource itself
public function authorizedToAttachAny(Request $request, $model)
    {
        return false; // to-do here
    }
```
